### PR TITLE
[ZEPPELIN-212] Multiple paragraph results

### PR DESCRIPTION
**Pull request by Leemoonsoo**
### What is this PR for?
Currently a paragraph can display only a single type of result.
For example
```
print("""
%text textout
%html htmlout
""")
```

will display only last display system detected, "htmlout".

This pr implements multiple results supports, so not only the last, but also all the display systems in a paragraph outputs can be displayed.

To do that, Note.json format need to be changed. from

```
   paragraph : [
      {
          result : {
              code: "",
              type: "",
              msg: ""            
          },
          config : {
             graph : {},
             ...
          },
          ...
      },
      ...
   ],
   ...
```

to

```
   paragraph : [
      {
          results : {
             {
                code: "",
                msg: [
                   {
                        type: "",
                        data: ""
                    },
                    {
                        type: "",
                        data: ""
                     },
                     ...
                ]
          },
          config : {
             results : [
                {
                    graph : {},
                },
                {
                    graph : {},
                },
                ...
             ]
          },
          ...
      },
      ...
   ],
   ...
```

### What type of PR is it?
Improvement

### Todos
* [x] - Make InterpreterResult and InterpreterOutput support multiple display system
* [x] - Automatic migration old format to new format
* [x] - Render multiple results in front-end
* [x] - Take care Importing old notebook format
* [x] - Make helium framework support multiple results

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-212

### How should this be tested?

run following code in a single spark paragraph.
```
%spark
// default output is text
(1 to 3).foreach{i=>
  println(new java.util.Date())
  Thread.sleep(1000)
}

// print something in html
println(s"""%html <h3><font style="color:blue">Some HTML</font></h3>""")

// create table
println(s"""%table key\tvalue
sun\t100
moon\t200

""")

// display text again
println("""%text""")
(1 to 3).foreach{i=>
  println(new java.util.Date())
  Thread.sleep(1000)
}

// another table
Thread.sleep(1000)
println(s"""%table key\tvalue
apple\t100
banana\t200
""")
```

### Screenshots (if appropriate)
multipleout

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? yes
* Does this needs documentation? yes

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
I think this PR is ready to be reviewed.
While this PR includes quite a lot of changes i'm highlighting some key changes.

1. Display multiple types of output in a single paragraph.

   Previously, if output includes multiple display system directive (e.g. %html, %text, etc) only the last one was effective and all the previous output was ignored. After the change, a paragraph sequentially display all display system directive in the output result.

   To do that, on the front-end side,
   Created ResultCtrl controller under `zeppelin-web/src/app/notebook/paragraph/result`
   and moved paragraph result rendering related scripts/templates under this directory.

    /notebook/paragraph/paragraph-graph.html -> /notebook/paragraph/result/result-graph.html
    /notebook/paragraph/paragraph-pivot.html -> /notebook/paragraph/result/result-pivot.html
    /notebook/paragraph/paragraph-result.html -> /notebook/paragraph/result/result-results.html
    /notebook/paragraph/paragraph-chart-selector.html -> /notebook/paragraph/result/result-chart-selector.html

    on the back-end side,
    Created `zeppelin-interpreter/interpreter/InterpreterResultMessageOutput.java` to take care of single output stream and existing `zeppelin-interpreter/interpreter/InterpreterResultMessageOutput.java` changed to take care of multiple outputs through `InterpreterResultMessageOutput`

2. Compatibility between old, new notebook format.

   note.json format is changed to store multiple outputs in the paragraph.

   To read note.json created by previous versions of Zeppelin, [Notebook.convertFromSingleResultToMultipleResultsFormat()](https://github.com/apache/zeppelin/blob/7ab6679a87ed75f176ffa61d8cf6c73f4241c528/zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java#L385) will automatically convert format when note.json is loaded in Zeppelin.

   If old version of Zeppelin reads notebook created by newer version, everything will be load normally except for output. Output will be empty.

   In detail, note.json is changed from

   ```
   paragraph : [
      {
          result : {
              code: "",
              type: "",
              msg: ""            
          },
          config : {
             graph : {},
             ...
          },
          ...
      },
      ...
   ],
   ...
   ```

   to

   ```
   paragraph : [
      {
          results : {
             {
                code: "",
                msg: [
                   {
                        type: "",
                        data: ""
                    },
                    {
                        type: "",
                        data: ""
                     },
                     ...
                ]
          },
          config : {
             results : [
                {
                    graph : {},
                },
                {
                    graph : {},
                },
                ...
             ]
          },
          ...
      },
      ...
   ],
   ...
   ```

3. Every other changes are trivial

   Most changes are trivial change due to change of InterpreterOutput and InterpreterResult api changes.

**Comment by agoodm** [comment]
Very nice @Leemoonsoo 👍 

I have tested this with the matplotlib tutorial notebook just to see if [ZEPPELIN-1638](https://issues.apache.org/jira/browse/ZEPPELIN-1638) is also resolved by this, and everything checks out!

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java [hunk 1]
```diff
@@ -383,6 +382,53 @@ public Note getNoteByRevision(String noteId, String revisionId, AuthenticationIn
...
+                results.add(res);
+              } else {
+                results.add(new HashMap<>());
+              }
+            }
+            config.put("results", results);
+          }
+        }
+      } catch (Exception e) {
+        logger.error("Conversion failure", e);
```
Comment (khalidhuseynov) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: would it be able to continue if the conversion has failed?

**Comment by khalidhuseynov** [comment]
tried locally and works as expected, LGTM!

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java (hunk 1 above)
Comment (khalidhuseynov) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: I see, so basically it handles old result format notes on frontend as well

Comment (Leemoonsoo) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: Front end will display note without output, incase of conversion failure.

**Comment by Leemoonsoo** [comment]
Thanks @agoodm @khalidhuseynov for the review!
Merge to master if there're no more comments.

**Comment by agoodm** [comment]
@Leemoonsoo FYI, #1632 added another set of tests. The interpreter result format for these should be changed as well.

**Comment by Leemoonsoo** [comment]
@agoodm merged master and addressed tests added from #1632.

CI failure is not related. Merge to master.
image

**Pull request closed without merging**